daphne/src/testing.rs

```diff
@@ -341,10 +348,66 @@ impl DapLeader<BearerToken> for MockAggregator {
 
     async fn get_reports(
         &self,
-        _task_id: &Id,
-        _selector: &Self::ReportSelector,
+        task_id: &Id,
+        selector: &Self::ReportSelector,
     ) -> Result<Vec<Report>, DapError> {
-        unreachable!("not implemented");
+        let task_config = self
+            .get_task_config_for(task_id)
+            .ok_or_else(|| DapError::fatal("no task found"))?;
+
+        // Calculate batch interval.
+        let batch_interval = if let Some(ref interval) = selector.batch_info {
+            if !interval.is_valid_for(task_config) {
+                return Err(DapError::fatal("invalid batch interval"));
+            }
+            interval.clone()
+        } else {
+            task_config.current_batch_window(Utc::now().timestamp() as u64)
+        };
+
+        // Do some checks on batch interval.
+        if batch_interval.start > Utc::now().timestamp() as u64 {
+            return Err(DapError::fatal(
+                "invalid batch interval: start time cannot be in the future",
+            ));
+        }
+
+        // Lock report_store.
+        let agg_rate = selector.agg_rate as usize;
+        let mut reports = Vec::with_capacity(agg_rate);
+        let mut report_store_mutex_guard = self
+            .report_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let report_store = report_store_mutex_guard.deref_mut();
+
+        // Fetch reports.
+        let mut window = batch_interval.start;
+        // TODO(nakatsuka-y) Should the batch window be inclusive?
+        while window < batch_interval.end() {
+            let num_reports_remaining = agg_rate - reports.len();
+            let bucket_info = BucketInfo {
+                task_id: task_id.clone(),
```

Comment:
We can easily avoid this clone by moving `bucket_info` out of the loop and updating `bucket_info.window` as needed.

Reply:
Great catch.